WIP: Heat transfer transient adjoint

jamiebramwell opened the pull request:
Proposed workflow:

1. Implement a parameterized dynamic heat conduction physics module.
2. Complete a full forward simulation.
3. Obtain previous states using the new `previousTemperature(...)` method.
4. Calculate a QoI based on the temperature history.
5. Compute a possibly time-varying adjoint load (d QOI/d state)
6. Call `reverseAdjointTimestep` to perform one backward Euler step for the adjoint solve using the timestep from the forward pass.
7. Call `computeTimestepSensitivity` to calculate the sensitivity (d QOI/d parameter) for the current timestep.

Some notes:

- Currently hardcoded to only use backward Euler time integrators.
- It reads the previous temperature values from disk at every timestep (no checkpointing).
- It changed the quasi-static function signature from `solveAdjoint` to `reverseAdjointTimestep` to be consistent with the forward pass.
- It updates Axom to include the `axom::MFEMSidreDataCollection` bugfix needed to correctly read an `mfem::ParGridFunction` from disk.

jamiebramwell committed:
feature: First work at adding a transient adjoint thermal solver. It is currently restricted to Backward Euler timestepping only. It reads previous state vectors from disk at every step.

jamiebramwell committed:
feature: Add accumulation of sensitivities for non-homogenous boundary conditions

jamiebramwell committed:
Add dynamic thermal adjoint test, remove debug statement, and add method for querying ODE solver type

jamiebramwell committed:
Merge branch 'develop' into feature/bramwell/transient_adjoint

tupek2 commented on src/serac/physics/state/state_manager.cpp:
```diff
@@ -92,6 +92,28 @@ double StateManager::newDataCollection(const std::string& name, const std::optio
+  previous_datacoll.Load(cycle_to_load);
+
+  for (auto state : states_to_load) {
+    SLIC_ERROR_ROOT_IF(collectionID(&states_to_load.begin()->get().mesh()) != mesh_name,
```
Should this be state->get().mesh()?  Currently its just accesses .begin() for every iteration, so its not really checking anything.

tepperly commented on src/serac/numerics/odes.hpp:
```diff
@@ -163,6 +163,8 @@ class SecondOrderODE : public mfem::SecondOrderTimeDependentOperator {
    */
   const State& GetState() { return state_; }
 
+  TimestepMethod GetTimestepper() { return timestepper_; }
```
Doxygen?

tepperly commented on src/serac/physics/base_physics.cpp:
```diff
@@ -45,10 +47,14 @@ void BasePhysics::setTime(const double time) { time_ = time; }
 
 double BasePhysics::time() const { return time_; }
 
+double BasePhysics::adjointTime() const { return adjoint_time_; }
```
This is the time during the adjoint stepping?

jamiebramwell replied:
yes - I thought it was better to have this be a separate field vs. the forward pass, but I am open to suggestions.

tupek2 replied:
Having them be different implies they can sometimes go out of sync?  What is the situation where the adjointTime and time would return different values?

tepperly commented on src/serac/physics/base_physics.cpp:
```diff
@@ -45,10 +47,14 @@ void BasePhysics::setTime(const double time) { time_ = time; }
 
 int BasePhysics::cycle() const { return cycle_; }
 
+int BasePhysics::adjointCycle() const { return adjoint_cycle_; }
```
What's the adjoint cycle concept? Is it an iteration count or what?

jamiebramwell replied:
It's the iteration count on the backward pass. I added some extra docs, but let me know if this isn't clear.

tepperly commented on src/serac/physics/heat_transfer.hpp:
```diff
@@ -650,39 +646,179 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
+                       "number of forward timesteps");
+
+    bool is_first_adjoint_timestep = (adjoint_cycle_ == -1);
+    bool is_last_adjoint_timestep  = (adjoint_cycle_ == 1);
```
Presumably these are const

tepperly reviewed (commented):
Is the end always defined by the final time? LiDO's API can potentially support other ways of ending. This could be managed in the LiDO wrapper.

jamiebramwell committed:
doc: Add doxygen

jamiebramwell committed:
Merge branch 'feature/bramwell/transient_adjoint' of github.com:LLNL/serac into feature/bramwell/transient_adjoint

jamiebramwell committed:
Merge branch 'develop' of github.com:LLNL/serac into feature/bramwell/transient_adjoint

jamiebramwell committed:
feature: Unify the interface for forward and backward (adjoint/sensitivity) ode evolution steps. This means that sensitivities are now provided incrementally - it is up to a user to integrate the time-varying d_QOI_d_sensitivity themselves.

jamiebramwell commented:
I changed the interface to reflect our discussion yesterday. The sensitivities are no longer accumulated internally. This means only the sensitivity increment is returned to a user for each timestep. I also added a field for the timestep size to be consistent with the forward pass. Let me know what you think!

jamiebramwell commented:
Closed in favor of https://github.com/LLNL/serac/pull/998
